### PR TITLE
Integration of sugar moiety removal functionality

### PR DESCRIPTION
**Pull request by JonasSchaub**
Reference issue: https://github.com/cdk/cdk/issues/854

### CODE REVIEW

**Comment by johnmay** [comment]
Thanks for the updates, I'm quite busy this week/early next but will do an in depth check end of next week. Overall from first glance it looks good and should be OK to merge reasonably quickly - I added one comment about the IPseudoAtom - but Symbol=R is also OK.

**Comment by johnmay** [comment]
This one I am less fussy about but some bits are similar:
- required: use the CDK logging tool rather than the JDK Logger
- required: remove all cloning
- required: remove throws of runtime exceptions
- optional: terser variable names, 'anBlah', 'aBlah', 'tmpBlah' I just really don't like that as a style.

**Comment by JonasSchaub** [comment]
@johnmay sorry for dropping the ball on this here; I will make the requested changes soon! Thanks for the initial review.

**Comment by johnmay** [comment]
Should just be to merge PR I sent you?

**Comment by JonasSchaub** [comment]
The PR you sent me is for the functional groups functionality. Good that you mention that because I haven't seen it before (incorrect "watch" settings on my fork, I'm very very sorry). Will get to that as well as soon as possible!

**Comment by johnmay** [comment]
Oh right sorry yes, moving house next week so things are all going 100kph ATM.

**Comment by JonasSchaub** [comment]
Good luck with that, hope it's going well!

**Comment by JonasSchaub** [comment]
@johnmay, I have made some adjustments according to your initial review:

- switched from JDK logger to CDK logging tool
- removed all cloning (I replaced it with a basic copy routine in one place, the copy is only needed there temporarily for testing whether a moiety is terminal or not)
- almost completely removed the throwing of runtime exceptions
- refactored variable names
- additionally: removed input restriction regarding unconnected structures
- additionally: simplified the API in some places
- additionally: added an example usage in the class doc

What do you think? Would this be ready to merge now?

**Comment by johnmay** [comment]
I will find time this week and take a look but all sounds reasonable

File: misc/extra/pom.xml [hunk 1]
```diff
@@ -72,6 +72,16 @@
...
             <version>${project.parent.version}</version>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>cdk-smarts</artifactId>
+            <version>${project.parent.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>cdk-silent</artifactId>
```
Comment (johnmay) [inline misc/extra/pom.xml:82]: Is there a way to remove this dependency - note for the QueryChemObjects you can just pass null for the builder.

**Review by johnmay — COMMENTED** [review]
Just one minor thing

File: misc/extra/pom.xml (hunk 1 above)
Comment (JonasSchaub) [reply misc/extra/pom.xml:82]: Hi @johnmay , thanks for reviewing!
I need the SilentChemObjectBuilder in the tests to parse SMILES and be able to add implicit hydrogens. Should I maybe add the test scope in the pom? Or I could switch to DefaultChemObjectBuilder which would apparently not require this extra dependency.

Comment (johnmay) [reply misc/extra/pom.xml:82]: Both OK but I would try and add ``<scope>test</scope>``

Comment (JonasSchaub) [reply misc/extra/pom.xml:82]: Added test scope in the new commit. Would it be ready to merge now?

**Pull request merged**